src/docs/timestamp-prepare.dox

```diff
@@ -49,6 +49,24 @@ transactions are higher-level MongoDB operations, requiring cluster-level
 consensus on durability. Applications without similar requirements for prepared
 transactions should set the durable and commit timestamps to the same time.
 
+When a transaction has a durable timestamp later than its commit timestamp,
+reading its writes in a second transaction and then committing such that the
+second transaction becomes durable before the first can produce data
+inconsistency. To help prevent this, reads between the commit and durable
+timestamp will fail with ::WT_ROLLBACK if the global stable timestamp has not
```

Comment:
As per the recent conversation with @keithbostic, MongoDB read operations don't expect a WT_ROLLBACK error. MongoDB may not hit this scenario, but in case it happens can it cause some problem? How about returning a WT_NOTFOUND instead if this is the only update for the key or returning a historical version?

Based on this change, the entire approach can change, @keithbostic can you provide your views on the same?

Reply:
Interesting... I'd prefer to go with `rollback` if it's possible. My concern is that `notfound` could be interpreted by the caller as the row physically isn't there which I think could itself lead to interesting and hard-to-debug failures.

I agree the list of failures in the patch build was unexpected. I guess my plan is to make this our next SERVER ticket once the WT-8745/SERVER-63166 pair is resolved, and understand the failures. Hopefully they're fixable, but if not, maybe `notfound` is our fallback position. I'm going to run a patch build with `notfound`, and see if that makes a difference.

Let me know if I'm misunderstanding the implications of returning notfound, please.

Reply:
Returning the previous version (or WT_NOTFOUND if there is no previous version) is mostly equivalent to rounding the commit timestamp up to the durable timestamp. I suspect that's not what we want, since there's then no point distinguishing the commit timestamp from the durable timestamp.

(I say "mostly equivalent" because if this applies only when stable hasn't moved up to the durable timestamp yet, then it's equivalent to that only before that point. But changing what you see from one value to another when reading in that interval, depending on where stable is, seems like a bad plan also.)

Always returning WT_NOTFOUND in that interval seems bad because it's providing wrong data rather than no data, so I'd rather not go that way if possible.

There's another completely different approach, which is that what we're doing here is trying to avoid generating dependent transactions. But dependent transactions aren't catastrophic and this is a minor manifestation of them; we could just track the dependence. I'll make a top-level post about this so it doesn't get buried.

Reply:
Thanks for the details @keithbostic and @sauclovian-wt. 
1. Returning WT_ROLLBACK has problems with MDB as they don't handle it. If MDB supports handling WT_ROLLBACK for all the read operations it shouldn't be a problem.
2. Returning WT_NOTFOUND has problems as described by David.

I am thinking of another return type WT_PREPARE_CONFLICT instead of either WT_ROLLBACK or WT_NOTFOUND as this issue occurs only with prepared transactions. I believe that MDB should have the code to handle the WT_PREPATE_CONFLICT error for reading operations.

Reply:
That's an interesting idea -- thank you, @kommiharibabu.

Let's run a patch check, and see if that passes the MDB server tests.

Reply:
Here's the patch run: https://evergreen.mongodb.com/version/62014cbc2a60ed3347152d81?redirect_spruce_users=true